// @stoplight/spectral-core/dist/ruleset/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rule = exports.Ruleset = exports.createRulesetFunction = exports.getDiagnosticSeverity = exports.RulesetValidationError = exports.assertValidRuleset = void 0;
var validation_1 = require("./validation");
Object.defineProperty(exports, "assertValidRuleset", { enumerable: true, get: function () { return validation_1.assertValidRuleset; } });
Object.defineProperty(exports, "RulesetValidationError", { enumerable: true, get: function () { return validation_1.RulesetValidationError; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getDiagnosticSeverity", { enumerable: true, get: function () { return utils_1.getDiagnosticSeverity; } });
var rulesetFunction_1 = require("./rulesetFunction");
Object.defineProperty(exports, "createRulesetFunction", { enumerable: true, get: function () { return rulesetFunction_1.createRulesetFunction; } });
var ruleset_1 = require("./ruleset");
Object.defineProperty(exports, "Ruleset", { enumerable: true, get: function () { return ruleset_1.Ruleset; } });
var rule_1 = require("./rule");
Object.defineProperty(exports, "Rule", { enumerable: true, get: function () { return rule_1.Rule; } });
